### ISSUE TITLE
Doubt in running evaluate_IL.py

### ISSUE THREAD

**Issue by sahilm1992** [ds4dm/branch-search-trees#6]
Hi,
I was trying to use the model provided in trained-models.

I had a few doubts:
for instance:

for map18 for seeds 0 1 2 3 4
I am getting the following number of nodes:
653, 747, 739, 519, 681

However, the paper reports shifted geometric mean of  ~575.

Could the results vary too much ?

**Comment by vis-opt** [comment]
In general this is a well-known phenomenon in operations research. For reference see ["Performance Variability in Mixed-Integer Programming" ](https://pubsonline.informs.org/doi/abs/10.1287/educ.2013.0112)by Andrea Lodi and Andrea Tramontani [ALAT].

It is known that fully deterministic computational algorithms are not fully reproducible, but why is the performance variability seemingly worse for us? This is a valid question to ask. In deep convolutional networks, it's not unusual to see a difference of less than 1% between different computational back-ends (e.g. using cuDNN vs. not using cuDNN). Why is it so much more drastic here in branching? 

Here is my best answer to your question of "Could the results vary too much ?"

My short answer is yes. I do believe the results can vary quite a bit. I try to give a longer answer below. My goal is not to convince you, but just share the way I view it. Having doubts is good and healthy, so if you are not convinced it's best you do more exploring. 

The issue lies in the nature of the problems you are trying to solve. In vision, images satisfy a local smoothness: by looking at one patch of the image, you can reasonably infer what other local patches are going to look like. This is just my own speculation, but this local smoothness prior manifests itself in the optimization landscape. Namely, for supervised classification, the optimization landscape consists of relatively "flat" local minima. These flatter regions have a correspondingly larger tolerance for variations (think about a function y = f(x) with a flat region around x_0, the y-values do not change a lot when you perturb around x_0) and tend to be robust and generalize well. So, using different hardware platforms, software back-ends, has a much smaller effect than what has been demonstrated in MIP solvers. 

We can analogously consider the "MIP solver landscape". One can argue that there is no such analogous local smoothness. For example, it is well known that in branching, it's possible to make a single bad decision and exponentially increase the size of your tree. Well, why would the branching policy ever make such a bad decision? Unlike supervised classification, there is actually no ground truth in branching. All we have are expertly-crafted heuristics that benchmark exceptionally well, but they are nonetheless heuristics. We do not have a theoretically optimal way to score the candidate branching variables. All heuristics have blind-spots or what behavioral psychologists call "cognitive biases." 

Now remember that MIP solving is inherently sequential with even "easy" problems taking hundreds of nodes to solve. Pessimistically, you can view the long MIP solver sequences as more and more opportunities for deterministic floating point errors ("imperfect tie breaking" from [ALAT] above as an example) to accumulate. It is then within the realm of possibility to end-up in a "blind spot" of the learned policy. This blind-spot is where the heuristic candidate scoring formula is high for a "bad" branching variable. Once a catastrophic branching decision is made, the size of the tree grows. 

Obviously there is a lot more to this story. It is arguably one of the biggest mysteries right now in all of computational complexity theory. But I hope this can at least shed some light on why I believe performance can vary quite a bit. 

More practically, it might be useful if you could share the setup you used to evaluate the pre-trained models? As in what software, what packages, hardware? In our paper we try to make reproducibility a bit less of a headache by sharing the recipe we used to generate our Singularity image. That's only half the battle though since everyone's hardware is going to differ. 

For whatever it is worth, we shared all the evaluation pickle files, so maybe you can diagnose the differences between your evaluation and our evaluation. Perhaps the `max_depth` differs, etc.

**Comment by sahilm1992** [comment]
Thank you very much for sharing your deep insights on this topic.
 I highly appreciate that :)

**Comment by Zishanna** [comment]
Hi，
Your deep insights also inspired me a lot. I have a related question. I used the same seed to run `evaluate_IL.py` several times, and got the different number of nodes every time.  This has bothered me for a long time. Could you explain this for me ?
